tests/src/test/scala/whisk/core/containerpool/test/ContainerProxyTests.scala

```diff
@@ -0,0 +1,458 @@
+        run(machine, Paused)
+
+        awaitAssert {
+            container.initializeCount shouldBe 1
+            container.runCount shouldBe 2
+            container.logsCount shouldBe 2
+            container.haltCount shouldBe 1
+            container.resumeCount shouldBe 1
+            ack.verify(message.transid, *).repeat(2)
+            store.verify(message.transid, *).repeat(2)
+        }
+    }
+
+    it should "successfully run on an uninitialized container" in within(timeout) {
+        val container = new TestContainer
+        def factory(tid: TransactionId, name: String, exec: Exec, memoryLimit: ByteSize) = Future.successful(container)
+
+        val machine = childActorOf(ContainerProxy.props(factory, ack, store))
+        registerCallback(machine)
+        run(machine, Uninitialized)
+
+        awaitAssert {
+            container.initializeCount shouldBe 1
+            container.runCount shouldBe 1
+            container.logsCount shouldBe 1
+            ack.verify(message.transid, *).repeat(1)
+            store.verify(message.transid, *).repeat(1)
+        }
+    }
+
+    /*
+     * ERROR CASES
+     */
+    it should "complete the transaction and abort if container creation fails" in within(timeout) {
+        val container = new TestContainer
+        def factory(tid: TransactionId, name: String, exec: Exec, memoryLimit: ByteSize) = Future.failed(new Exception())
+
+        val machine = childActorOf(ContainerProxy.props(factory, ack, store))
+        registerCallback(machine)
+        machine ! Run(action, message)
+        expectMsg(Transition(machine, Uninitialized, Running))
+        expectMsg(ContainerRemoved)
+
+        awaitAssert {
+            container.initializeCount shouldBe 0
+            container.runCount shouldBe 0
+            container.logsCount shouldBe 0 // gather no logs
+            container.destroyCount shouldBe 0 // no destroying possible as no container could be obtained
+            ack.verify(message.transid, *).repeat(1)
+            store.verify(message.transid, *).repeat(1)
+        }
+    }
+
+    it should "complete the transaction and destroy the container on a failed init" in within(timeout) {
+        val container = new TestContainer {
+            override def initialize(initializer: JsObject, timeout: FiniteDuration)(implicit transid: TransactionId): Future[Interval] = {
+                initializeCount += 1
+                Future.failed(InitializationError(Interval.zero, ActivationResponse.applicationError("boom")))
+            }
+        }
+        def factory(tid: TransactionId, name: String, exec: Exec, memoryLimit: ByteSize) = Future.successful(container)
+
+        val machine = childActorOf(ContainerProxy.props(factory, ack, store))
+        registerCallback(machine)
+        machine ! Run(action, message)
+        expectMsg(Transition(machine, Uninitialized, Running))
+        expectMsg(ContainerRemoved)
+        expectMsg(Transition(machine, Running, Removing))
```

Comment:
The `ContainerRemoved` message is sent long before the container is really removed. It's sent in the moment the container decides to remove itself, and thus tells its parent it's gone already (as the parent should treat it like its gone).

You're right, from this perspective it's confusing. I'll have a look.